### PR TITLE
OHRM 251: Modified the logic of instance identifier creation to make the identifier unique for particular instance 

### PR DESCRIPTION
**Pull request by Bashitha**

### CODE REVIEW

**Comment by RajithaKumara** [comment]
Previously there was a reason to send registration data, before run installation steps (ex: DB creation, Initial data insertion, Config file creation, etc). We had a need to track the installation completion status. When the user confirms the installation, the installer will send registration data before DB creation. Then the Penguin server knows a particular user start an installation. After successfully complete the installation and when the user successfully signin into the OrangeHRM OS instance, then the system will send successful status to the Penguin server. Then Penguin server knows the particular user successfully installed the OrangeHRM OS. Any conflicts? Ref PR: #349

**Comment by Bashitha** [comment]
Thanks Rajitha for the reference. Thanks for letting me know that there was a requirement of sending the registration data to the penguin server before installing the system. So when the registration data is sending after the installation? I couldn't figure out  an occasion of sending data after the installation.  Is it after login into the system? I couldn't checkout that instance. So basically are there 2 occasions of sending registration data to the penguin server for one particular installation?
Basically Kavitha had a issue of duplicating the same registration data in the penguin server side. That's why I started looking into this issue. When debugging I figured out that there was a possibility of cancelling the installation after hitting the install button, but in that instance also registration data was sent to the penguin server side. After discussing with Nuwan we decided there is no need of sending registration data before installing the system.

File: installer/SystemConfiguration.php [hunk 1]
```diff
@@ -144,45 +144,74 @@ public function createAdminUser($userName, $password) {
...
-     * @param $randomNumber
+     * @param $adminFirstName
+     * @param $adminLastName
+     * @param $host
+     * @param $country
+     * @param $ohrmVersion
      */
-    public function setInstanceIdentifier($organizationName, $email, $randomNumber) {
-        $instanceIdentifier = $organizationName . '_' . $email . '_' . date('Y-m-d') . $randomNumber;
+    public function setInstanceIdentifier($organizationName, $email, $adminFirstName, $adminLastName, $host, $country, $ohrmVersion) {
```
Comment (samanthajayasinghe) [inline installer/SystemConfiguration.php]: pls use as multiline function for readability

File: installer/SystemConfiguration.php [hunk 2]
```diff
@@ -144,45 +144,74 @@ public function createAdminUser($userName, $password) {
...
+     * @param $host
+     * @param $country
+     * @param $ohrmVersion
      * @return string
      */
-    public function createInstanceIdentifier($organizationName, $organizationEmail, $createdDate, $randomNumber) {
-        return base64_encode($organizationName . '_' . $organizationEmail . '_' . $createdDate . $randomNumber);
+    public function createInstanceIdentifier($organizationName, $email, $adminFirstName, $adminLastName, $host, $country, $ohrmVersion) {
+        if(is_null($host))
+        {
```
Comment (samanthajayasinghe) [inline installer/SystemConfiguration.php]: This is against PSR2 coding standards 
https://www.php-fig.org/psr/psr-2/#5-control-structures

File: installer/SystemConfiguration.php [hunk 3]
```diff
@@ -144,45 +144,74 @@ public function createAdminUser($userName, $password) {
...
+     * @param $adminFirstName
+     * @param $adminLastName
+     * @param $host
+     * @param $country
+     * @param $ohrmVersion
      * @return string
      */
-    public function createInstanceIdentifier($organizationName, $organizationEmail, $createdDate, $randomNumber) {
-        return base64_encode($organizationName . '_' . $organizationEmail . '_' . $createdDate . $randomNumber);
+    public function createInstanceIdentifier($organizationName, $email, $adminFirstName, $adminLastName, $host, $country, $ohrmVersion) {
```
Comment (samanthajayasinghe) [inline installer/SystemConfiguration.php]: better to use PSR2 standards multiline 
https://www.php-fig.org/psr/psr-2/#6-closures

File: installer/SystemConfiguration.php [hunk 4]
```diff
@@ -144,45 +144,74 @@ public function createAdminUser($userName, $password) {
...
+     * @param $email
+     * @param $adminFirstName
+     * @param $adminLastName
+     * @param $host
+     * @param $country
+     * @param $ohrmVersion
      * @return string
      */
-    public function createInstanceIdentifierChecksum($organizationName, $organizationEmail, $createdDate, $randomNumber) {
+    public function createInstanceIdentifierChecksum($organizationName, $email, $adminFirstName, $adminLastName, $host, $country, $ohrmVersion) {
```
Comment (samanthajayasinghe) [inline installer/SystemConfiguration.php]: standard multiline

File: installer/SystemConfiguration.php [hunk 5]
```diff
@@ -191,15 +220,28 @@ public function createInstanceIdentifierChecksum($organizationName, $organizatio
     /**
      * Set the instance identifier checksum value
      * @param $organizationName
-     * @param $organizationEmail
-     * @param $randomNumber
+     * @param $email
```
Comment (samanthajayasinghe) [inline installer/SystemConfiguration.php]: comment type is missing

File: upgrader/lib/UpgradeSystemConfiguration.php [hunk 6]
```diff
@@ -243,14 +241,11 @@ public function hasSetInstanceIdentifier() {
...
+     * Set the instance identifier checksum     *
      */
-    public function setInstanceIdentifierChecksum($instanceIdentifierChecksum) {
-        $query = "INSERT INTO `hs_hr_config` (`key`, `value`) VALUES (?, ?)";
-        $dbConnection = $this->createDbConnection();
-        $statement = $dbConnection->prepare($query);
-        $statement->execute(array(self::KEY_INSTANCE_IDENTIFIER_CHECKSUM, $instanceIdentifierChecksum));
+    public function setInstanceIdentifierChecksum() {
+        $this->sysConf = $this->getSystemConfiguration();
+        $this->sysConf->setInstanceIdentifierChecksum($_SESSION['defUser']['organizationName'], $_SESSION['defUser']['organizationEmailAddress'], $_SESSION['defUser']['adminEmployeeFirstName'], $_SESSION['defUser']['adminEmployeeLastName'], $_SERVER['HTTP_HOST'], $_SESSION['defUser']['country'] , $this->sysConf->getOhrmVersion());
```
Comment (samanthajayasinghe) [inline upgrader/lib/UpgradeSystemConfiguration.php]: standard multiline

File: upgrader/lib/UpgradeSystemConfiguration.php [hunk 7]
```diff
@@ -210,11 +211,8 @@ private function getAdminEmployeeNumber() {
      * Set the instance identifier of upgraded instance
      */
     public function setInstanceIdentifier() {
-        $instanceIdentifier = $_SESSION['defUser']['organizationName'] . '_' . $_SESSION['defUser']['organizationEmailAddress'] . '_' . date('Y-m-d') . $_SESSION['defUser']['randomNumber'];
-        $query = "INSERT INTO `hs_hr_config` (`key`, `value`) VALUES (?, ?)";
-        $dbConnection = $this->createDbConnection();
-        $statement = $dbConnection->prepare($query);
-        $statement->execute(array(self::KEY_INSTANCE_IDENTIFIER, base64_encode($instanceIdentifier)));
+        $this->sysConf = $this->getSystemConfiguration();
+        $this->sysConf->setInstanceIdentifier($_SESSION['defUser']['organizationName'], $_SESSION['defUser']['organizationEmailAddress'], $_SESSION['defUser']['adminEmployeeFirstName'], $_SESSION['defUser']['adminEmployeeLastName'], $_SERVER['HTTP_HOST'], $_SESSION['defUser']['country'] , $this->sysConf->getOhrmVersion());
```
Comment (samanthajayasinghe) [inline upgrader/lib/UpgradeSystemConfiguration.php]: standard multiline

File: installer/SystemConfiguration.php (hunk 5 above)
Comment (Bashitha) [reply installer/SystemConfiguration.php]: Didn't get this. Can you please explain more?

File: installer/SystemConfiguration.php [hunk 8]
```diff
@@ -144,45 +144,102 @@ public function createAdminUser($userName, $password) {
...
+        $country,
+        $ohrmVersion
+    ) {
+        if (is_null($host)) {
+            $host = '';
+        }
+        if (is_null($country)) {
+            $country = '';
+        }
+        return base64_encode($organizationName . '_' . $email . '_' . $adminFirstName . '_' . $adminLastName . '_' . $host . '_' . $country . '_' . $ohrmVersion);
```
Comment (samanthajayasinghe) [inline installer/SystemConfiguration.php]: This should be multiline

File: installer/SystemConfiguration.php (hunk 5 above)
Comment (samanthajayasinghe) [reply installer/SystemConfiguration.php]: I mean 
 @param string $email

File: upgrader/lib/UpgradeSystemConfiguration.php [hunk 9]
```diff
@@ -243,14 +248,18 @@ public function hasSetInstanceIdentifier() {
...
-     * @param $instanceIdentifierChecksum
+     * Set the instance identifier checksum     *
      */
-    public function setInstanceIdentifierChecksum($instanceIdentifierChecksum) {
-        $query = "INSERT INTO `hs_hr_config` (`key`, `value`) VALUES (?, ?)";
-        $dbConnection = $this->createDbConnection();
-        $statement = $dbConnection->prepare($query);
-        $statement->execute(array(self::KEY_INSTANCE_IDENTIFIER_CHECKSUM, $instanceIdentifierChecksum));
+    public function setInstanceIdentifierChecksum() {
+        $this->sysConf = $this->getSystemConfiguration();
```
Comment (samanthajayasinghe) [inline upgrader/lib/UpgradeSystemConfiguration.php]: is that spacing right with other params?

File: upgrader/lib/UpgradeSystemConfiguration.php [hunk 10]
```diff
@@ -277,4 +286,16 @@ public function hasSetInstanceIdentifierChecksum() {
...
     }
+
+    /**
+     * gets an instance of SystemConfiguration
+     * @return SystemConfiguration
+     */
+    private function getSystemConfiguration() {
+        if(is_null($this->sysConf)) {
+            $this -> sysConf = new SystemConfiguration();
+
```
Comment (samanthajayasinghe) [inline upgrader/lib/UpgradeSystemConfiguration.php]: extra line

File: upgrader/lib/UpgradeSystemConfiguration.php (hunk 9 above)
Comment (Bashitha) [reply upgrader/lib/UpgradeSystemConfiguration.php]: Yeah Samantha. This has 4 spaces for the indentation according to the PSR-2 code format
